zerver/migrations/0453_backfill_remote_realmauditlog_extradata_to_json_field.py

```diff
@@ -0,0 +1,86 @@
+    # We do not need to skip existing entries for other parts of backfilling because we have double-write implemented
+    # so that the backfilled value will still be consistent.
+    audit_log_model.objects.filter(
+        event_type=USER_FULL_NAME_CHANGED,
+        id__range=(id_lower_bound, id_upper_bound),
+        extra_data_json={},
+    ).update(extra_data_json=JSONObject(**{OLD_VALUE: "extra_data", NEW_VALUE: None}))
+
+    # A dict converted with str() will start with a open bracket followed by a single quote,
+    # as opposed to a JSON-encoded value, which will use a _double_ quote.
+    # We use this to filter out those entries with malformed extra_data to be handled later.
+    # This should only update rows with extra_data populated with orjson.dumps.
+    (
+        audit_log_model.objects.filter(
+            extra_data__isnull=False, id__range=(id_lower_bound, id_upper_bound)
+        )
+        .exclude(extra_data__startswith="{'")
+        .exclude(event_type=USER_FULL_NAME_CHANGED)
+        .update(extra_data_json=Cast("extra_data", output_field=JSONField()))
+    )
+
+    python_valued_audit_log_entries = audit_log_model.objects.filter(
+        extra_data__startswith="{'", id__range=(id_lower_bound, id_upper_bound)
+    )
+    for audit_log_entry in python_valued_audit_log_entries:
+        # extra_data for entries that store dict stringified with builtins.str() are converted
+        # back with ast.literal_eval for safety and efficiency.
+        # str()'d extra_data with the REALM_DISCOUNT_CHANGED event type is not handled by this
+        # migration. We expect that all such entries are manually converted beforehand or an
+        # error will occur during the migration, because ast.literal_eval does not allow the
+        # evaluation of Decimal.
+        audit_log_entry.extra_data_json = ast.literal_eval(audit_log_entry.extra_data)  # type: ignore[attr-defined] # The migration cannot depend on zerver.models, which contains the real type of the RealmAuditLog model, so it cannot be properly typed.
+    audit_log_model.objects.bulk_update(python_valued_audit_log_entries, fields=["extra_data_json"])
+
+
+def backfill_extra_data(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    audit_log_model = apps.get_model("zerver", "RealmAuditLog")
+    if not audit_log_model.objects.filter(extra_data__isnull=False).exists():
+        return
+
+    audit_log_entries = audit_log_model.objects.filter(extra_data__isnull=False)
+    id_lower_bound = audit_log_entries.earliest("id").id
+    id_upper_bound = audit_log_entries.latest("id").id
+    while id_lower_bound <= id_upper_bound:
+        do_bulk_backfill_extra_data(
+            audit_log_model, id_lower_bound, min(id_lower_bound + BATCH_SIZE, id_upper_bound)
+        )
+        id_lower_bound += BATCH_SIZE + 1
+
+
+class Migration(migrations.Migration):
```

Comment:
You need `atomic=False` whenever using a BATCH_SIZE style migration; otherwise the whole migration is a transaction.

(We might want to figure out a way to lint for this sort of thing).

Reply:
Looks like we are missing the `@transaction.atomic` decorator for some of the previous batched migrations (e.g.: `0380_userprofile_uuid_backfill`, `0386_fix_attachment_caches`). ` We can have some semgrep patterns for that.

Reply:
> Looks like we are missing the @transaction.atomic decorator for some of the previous batched migrations (e.g.: 0380_userprofile_uuid_backfill, 0386_fix_attachment_caches)
Nevermind, these migrations are irrelevant because they only do a single query per batch.